File: Data Structures & Algorithms/count-connected-components/submission-5.java
class Solution {
    public void dfs(int node, int[] visited, HashMap<Integer, ArrayList<Integer>> graph)
    {
        // if(visited[node] != 0)
        // {
        //     return false;
        // }
        visited[node] = 1;
        ArrayList<Integer> neighbors = graph.get(node);
        for(int i = 0; i < neighbors.size(); i++)
        {
            if(visited[neighbors.get(i)] == 0)
            {
                dfs(neighbors.get(i), visited, graph);
            }
        }
        // return true;
    }
    public int countComponents(int n, int[][] edges) {
        HashMap<Integer, ArrayList<Integer>> graph = new HashMap<>();
        for(int i = 0; i < n; i++)
        {
            graph.put(i, new ArrayList<Integer>());
        }
        for(int i = 0; i < edges.length; i++)
        {
            graph.get(edges[i][0]).add(edges[i][1]);
            graph.get(edges[i][1]).add(edges[i][0]);
        }

        int count = 0;
        int[] visited = new int[n];
        for(int i = 0; i < n; i++)
        {
            if(visited[i] == 0)
            {
                dfs(i, visited, graph);
                count++;
            }
            // if(dfs(i, visited, graph))
            // {
            //     count++;
            // }
        }

        return count;
    }
}

File: Data Structures & Algorithms/valid-tree/submission-5.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        d = defaultdict(list)
        for a, b in edges:
            d[a].append(b)
            d[b].append(a)

        visited = [0] * n
        has_cycle = False

        def dfs(i, parent):
            nonlocal has_cycle
            if visited[i] == 1:
                has_cycle = True
                return
            if visited[i] == 2:
                return

            visited[i] = 1
            for u in d[i]:
                if u != parent:
                    dfs(u, i)
            visited[i] = 2
        
        dfs(0, -1)
        return not has_cycle and len(edges) == n - 1